### PR TITLE
Added 'DatasourceCredentials' object

### PR DESCRIPTION
**Pull request by wbreza**
Introduced gateway credentials object with support for encrypting and decrypting credentials.  Using the same AsymmetricKeyEncryptionHelper that existed in DM.

### CODE REVIEW

File: sdk/PowerBI.Core/Cryptography/AsymmetricKeyEncryptionHelper.cs [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+using System.Xml.Serialization;
+
+namespace Microsoft.PowerBI.Core.Cryptography
+{
+    /// <summary>
+    /// Helper class for encrypting and decrypting with asymmetric keys
+    /// </summary>
+    public static class AsymmetricKeyEncryptionHelper
+    {
+        private const int SegmentLength = 85;
```
Comment (aluong) [inline sdk/PowerBI.Core/Cryptography/AsymmetricKeyEncryptionHelper.cs]: Does it make sense to make a options class and pass these configs in here?

Comment (aluong) [inline sdk/PowerBI.Core/Cryptography/AsymmetricKeyEncryptionHelper.cs]: Does it make sense to make a options class and pass these configs in here?

Comment (wbreza) [inline sdk/PowerBI.Core/Cryptography/AsymmetricKeyEncryptionHelper.cs]: These are the same fixed sizes that are used in DM.  Do we have a use case where we would need to support variable sizes?

File: sdk/PowerBI.Core/Cryptography/AsymmetricKeyEncryptionHelper.cs [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+        }
+
+        /// <summary>
+        /// Encrypts the text from the keys in the specified container
+        /// </summary>
+        /// <param name="plainText">The text to encrypt</param>
+        /// <param name="containerName">The key container name</param>
+        /// <returns></returns>
+        /// <exception cref="ArgumentNullException"></exception>
+        public static string Encrypt(string plainText, string containerName)
```
Comment (aluong) [inline sdk/PowerBI.Core/Cryptography/AsymmetricKeyEncryptionHelper.cs]: Is there no library that will just do this for us?

File: sdk/PowerBI.Core/Security/GatewayCredentials.cs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+            return part["value"]?.Value<string>();
+        }
+
+        /// <summary>
+        /// Encrypts the current instance
+        /// </summary>
+        /// <param name="containerName">The container name</param>
+        /// <returns>Base64 encoded encrypted credientals</returns>
+        public string Encrypt(string containerName)
```
Comment (aluong) [inline sdk/PowerBI.Core/Security/GatewayCredentials.cs]: Can we move the public methods before the private methods?

File: sdk/PowerBI.Core/Cryptography/AsymmetricKeyEncryptionHelper.cs (hunk 2 above)
Comment (wbreza) [inline sdk/PowerBI.Core/Cryptography/AsymmetricKeyEncryptionHelper.cs]: This is copied directly from the DM encryption method.  It appears it is non-standard to begin with and we'll need to conform to what DM expects

File: sdk/PowerBI.Core/Security/GatewayCredentials.cs [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+namespace Microsoft.PowerBI.Core.Security
+{
+    /// <summary>
+    /// The gateway credentials used to authticate to your datasources
+    /// </summary>
+    public class GatewayCredentials
+    {
+        /// <summary>
+        /// The username used to authorize to the gateway
```
Comment (aluong) [inline sdk/PowerBI.Core/Security/GatewayCredentials.cs]: Maybe we should GatewayCredentials be just the username and password. Then we have a static class that encrypts and decrypts the GatewayCredentials. I prefer the decoupling.

File: sdk/PowerBI.Core/Cryptography/AsymmetricKeyEncryptionHelper.cs [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+        /// <returns></returns>
+        /// <exception cref="ArgumentNullException"></exception>
+        public static AsymmetricKey ProduceAsymmetricKey(string containerName)
+        {
+            if (string.IsNullOrEmpty(containerName)) throw new ArgumentNullException(nameof(containerName));
+
+            var cspParams = new CspParameters { KeyContainerName = containerName };
+
+            // delete any old keys associated with the existing container
+            using (var rsa = new RSACryptoServiceProvider(cspParams))
```
Comment (aluong) [inline sdk/PowerBI.Core/Cryptography/AsymmetricKeyEncryptionHelper.cs]: Does this require us to run in admin mode?

File: sdk/PowerBI.Core/Security/GatewayCredentials.cs [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+using Microsoft.PowerBI.Core.Cryptography;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.PowerBI.Core.Security
+{
+    /// <summary>
+    /// The gateway credentials used to authticate to your datasources
+    /// </summary>
+    public class GatewayCredentials
```
Comment (aluong) [inline sdk/PowerBI.Core/Security/GatewayCredentials.cs]: Is this the credentials to authenticate with the Gateway or the Datasources behind the Gateway?

Comment (wbreza) [inline sdk/PowerBI.Core/Security/GatewayCredentials.cs]: It is the datasource credentials - Will rename class.

File: sdk/PowerBI.Core/Cryptography/AsymmetricKeyEncryptionHelper.cs (hunk 1 above)
Comment (aluong) [inline sdk/PowerBI.Core/Cryptography/AsymmetricKeyEncryptionHelper.cs]: Prob not now.

**Comment by aluong** [comment]
Looks good.

**Pull request closed without merging**